Runtime implementation of TypedDict extension

ilevkivskyi opened the pull request:
Here is the runtime implementation of ``TypedDict`` initially proposed in https://github.com/python/typing/pull/322, @davidfstr please take a look. You told you are interested only in keyword syntax, but here I add all three forms. I think it will be easy to add support for other forms in mypy (as it happened for ``NamedTuple``), I could do this later in a separate PR. The implementation is tested on both Python 2 and 3 but I didn't find how to add Python2 runtime tests in mypy.

I copy the description from original ``typing`` PR:

-------------------------------------------------------
Here is a simple (but quite flexible) implementation that supports three forms (iterable, keywords, class for 3.6+). In all forms type info for runtime introspection is accessible via ``__annotations__``, multiple inheritance is supported. Some examples:
```python
class Point1D(TypedDict):
    x: float
class Point2D(Point1D):
    y: float

Grid2D = TypedDict('Grid2D', [('x', int), ('y', int)])
Grid3D = TypedDict('Grid3D', {'x': int, 'y': int, 'z': int})
Info = TypedDict('Info', color=str, size=int)

class CoolPoint2D(Point2D, Info):
    pass

assert Point2D(x=1, y=2) == dict(x=1, y=2)
assert CoolPoint2D.__bases__ == (dict,)
assert CoolPoint2D.__annotations__ == {'x': float, 'y': float, 'size': int, 'color': str}
```
Everything is implemented to be pickleable and fast. I have measured and ``Point2D(x=1, y=2)`` is slower that ``dict(x=1, y=2)`` by only 3% to 11% depending on Python version.

ilevkivskyi commented on extensions/mypy_extensions.py:
```diff
@@ -8,15 +8,65 @@
-def TypedDict(typename, fields):
-    """TypedDict creates a dictionary type that expects all of its
+import sys
+from typing import _type_check  # type: ignore
```
I am using a non-public API here, but probably it is better than just copy the whole function here.

ilevkivskyi commented on mypy/test/testextensions.py:
```diff
@@ -0,0 +1,103 @@
+
+    @skipUnless(PY36, 'Python 3.6 required')
+    def test_class_syntax_usage(self):
+        self.assertEqual(LabelPoint2D.__annotations__, {'x': int, 'y': int, 'label': str})  # noqa
```
I added ``# noqa`` because flake8 complains about undefined name.

davidfstr commented on mypy/test/testextensions.py:
```diff
@@ -0,0 +1,103 @@
+
+    def test_basics_iterable_syntax(self):
+        # Check that two iterables allowed
+        Emp = TypedDict('Emp', [('name', str), ('id', int)])
```
I was under the impression that we weren't going to support the list of tuples syntax, based on https://github.com/python/typing/issues/28#issuecomment-249920175 and https://github.com/python/typing/issues/28#issuecomment-249949890 .

ilevkivskyi replied:
This is rather by accident. The implementation just takes an arbitrary iterable as an argument, e.g., a dictionary ``TypedDict('Employee', {'name': str, 'id': int})``. List of tuples etc. also work. Basically everything that could be given to a normal ``dict`` works. Of course we could prohibit everything apart from a real ``dict``, but I think that would be redundant. Anyway, I just removed this test, since I agree that it is unnecessary.

davidfstr commented on mypy/test/testextensions.py:
```diff
@@ -0,0 +1,103 @@
+        Emp = TypedDict('Emp', {'name': str, 'id': int})
+        self.assertIsSubclass(Emp, dict)
+        jim = Emp(name='Jim', id=1)
+        self.assertIsInstance(jim, Emp)
```
Shouldn't this check fail? I thought we were going to prevent isinstance() checks with TypedDicts since they would true for any dict type. Indeed in test_typeddict_errors() below that seems to be the case.

Actually in rereading your implementation it looks like the runtime type of Emp(...) is actually an Emp rather than a plain dict, as the TypedDict doc comment says. I would advocate that even if Emp itself has `__attributes__` and other nice introspection properties that instances still be plain dict objects. This is similar to how NewType works.

ilevkivskyi replied:
Yes, I remember I tried this, but it didn't give any visible speed-up. Current implementation is already slower than plain ``dict`` only by 10%, so that it is difficult to improve this. Anyway, probably you are right that it is better to make instances just ``dict``s for consistency with other types. I did this in a new commit (this unfortunately required a bit more "magic" to be compatible with both Python 2 and 3, note some other magic already present is only because of different class machinery in Python 2 and 3).

davidfstr commented on extensions/mypy_extensions.py:
```diff
@@ -8,15 +8,65 @@
+
+
+def _check_fails(cls, other):
+    if sys._getframe(1).f_globals['__name__'] not in ['abc', 'functools']:
```
(1) Is there a reason that these particular modules are special-cased so that they *do* allow isinstance checks?

(2) Perhaps the call to sys._getframe() should be guarded when using a non-CPython implementation that doesn't support sys._getframe()? I see this is done in `_TypedDictMeta.__new__` below.

davidfstr replied:
(1) Is there a reason that these particular modules are special-cased so that they *do* allow isinstance checks?

(2) Perhaps the call to sys._getframe() should be guarded when using a non-CPython implementation that doesn't support sys._getframe()? I see this is done in `_TypedDictMeta.__new__` below.

ilevkivskyi commented on extensions/mypy_extensions.py:
```diff
@@ -8,15 +8,65 @@
     """
-    def new_dict(*args, **kwargs):
-        return dict(*args, **kwargs)
+    def __new__(cls, _typename, fields=None, **kwargs):
```
Fixed.

ilevkivskyi commented:
@davidfstr Thank you for a thorough review! I implemented your comments in a new commit.

davidfstr commented on extensions/__init__.py:
```diff
@@ -0,0 +1 @@
+# This page intentioanlly left blank.
```
Nit: intentioanlly -> intentionally

davidfstr commented on mypy/test/testextensions.py:
```diff
@@ -0,0 +1,105 @@
+        Emp = TypedDict('Emp', {'name': str, 'id': int})
+        self.assertIsSubclass(Emp, dict)
+        jim = Emp(name='Jim', id=1)
+        self.assertIsInstance(jim, dict)
```
Suggest strengthen this check to:

```
self.assertTrue(type(jim) is dict)
```

Ditto for other occurrences below.

ilevkivskyi commented:
@davidfstr Thank you for another review! I implemented your comments (and added few more tests) in a new commit.

davidfstr commented:
Looks ready for squash-and-merge to me.

Any comments @gvanrossum or others with merge power?

gvanrossum commented:
I'll try to give it a quick look over and merge. I'm running behind with
stuff, as usual. :-(

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,76 @@
+            y: int
+            label: str
+
+    The latter syntax is only supported in Python 3.6+
```
Maybe add explicitly (as a reminder) that first two syntax forms works for Python 2(.7) and 3?

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,76 @@
+            _fields = kwargs
+        elif kwargs:
+            raise TypeError("Either list of fields or keywords"
+                            " can be provided to TypedDict, not both")
```
It's not a list but a dict, right? Anyway, I'd rewrite this error in the active tense, and without mentioning "fields" (which is questionable terminology here): "TypedDict takes either a dict or keyword arguments, but not both".

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,76 @@
-    new_dict.__name__ = typename
-    new_dict.__supertype__ = dict
-    return new_dict
+TypedDict = _TypedDictMeta('TypedDict', _TypedDict.__bases__, dict(_TypedDict.__dict__))
```
I'm guessing this awkward way of defining the class is to get around the different metaclass syntaxes for Python 2 and 3. But it makes my head explode trying to understand what's happening. I *think* that one of the problems is actually that `TypedDict` is itself a metaclass. (We see this when we realize that *instances* of TypedDict are really *classes* whose instances are dicts.) So maybe `_TypedDict` needs to be combined into `TypedDictMeta`? What tipped me off was that `_TypedDict.__new__` ends up not calling `object()` or `super().__new__()` but `cls.__class__()`, which is really a fancy way of spelling `TypedDictMeta()`.

gvanrossum commented on mypy/test/testextensions.py:
```diff
@@ -0,0 +1,115 @@
+except ImportError:
+    import collections as collections_abc  # type: ignore # PY32 and earlier
+from unittest import TestCase, main, skipUnless
+from extensions.mypy_extensions import TypedDict
```
This line should somehow be `from mypy_extensions import TypedDict`, see my comment on `__init__.py`.

gvanrossum commented on mypy/test/testextensions.py:
```diff
@@ -0,0 +1,115 @@
+            TypedDict('Hi', [('x', int)], y=int)
+
+    @skipUnless(PY36, 'Python 3.6 required')
+    def test_class_syntax_usage(self):
```
Can you add "py36" or something similar to the test case name?

ilevkivskyi commented on extensions/__init__.py:
```diff
@@ -0,0 +1 @@
+# This page intentionally left blank.
```
Situation was actually opposite here. I added this file because otherwise mypy was complaining. But now I see that I misunderstood the idea. I have removed this file.

ilevkivskyi commented:
@gvanrossum Thank you for a review! I implemented your comments in new commits. Also I have fixed an unguarded node look-up in ``process_typeddict_definition`` that was causing a crash. It looks like this crash is not related to https://github.com/python/mypy/issues/2535

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,76 @@
 
+import sys
+import types
+from typing import _type_check  # type: ignore
```
OK, it's fine to keep, but please add a comment to the code for posterity. (I also worry that before we know it this will become a de-facto public API -- I've seen others ask for the same functionality, e.g. in the issue asking to restore get_type_hint() for Python 2.)

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,78 @@
+        _fields = kwargs
+    elif kwargs:
+        raise TypeError("TypedDict takes either a dict or "
+                        "keyword arguments, but not both")
```
Can you break after the comma? (I worry about people grepping for the error message.)

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,78 @@
+            anns.update(base.__dict__.get('__annotations__', {}))
+        tp_dict.__annotations__ = anns
+        if name == 'TypedDict':
+            tp_dict.__new__ = types.MethodType(_typeddict_new, tp_dict)
```
AFAIK the signature of `MethodType` is undocumented. Why not just remove the unused `inst` arg from `_typeddict_new` and `_dict_new` and plunk the bare function into `tp_dict.__new__`? It seems to work.

ilevkivskyi replied:
> It seems to work.

This approach does not work in Python 2. If ``__new__`` returns an object of "wrong" class, then it looks like ``isinstance()`` is called on it and the latter is prohibited. Anyway, I have found a much simpler way: just insert ``__new__`` into class namespace before ``super().__new__`` i.e. ``type.__new__`` is called. The latter will do whatever is necessary for both Python 2 and 3.

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,78 @@
+
+def _check_fails(cls, other):
+    try:
+        if sys._getframe(1).f_globals['__name__'] not in ['abc', 'functools']:
```
Hm... If I comment out this line things still work, suggesting that it's not important that it returns False when invoked from those modules. I suppose this is from when this code was part of typing.py?

gvanrossum commented on extensions/mypy_extensions.py:
```diff
@@ -7,16 +7,78 @@
+        if name == 'TypedDict':
+            tp_dict.__new__ = types.MethodType(_typeddict_new, tp_dict)
+        else:
+            tp_dict.__new__ = types.MethodType(_dict_new, tp_dict)
```
Also, I finally remember why this is so complicated -- we must support both *calling* TypedDict and *subclassing* it. I like the new refactoring better, but maybe it's worth adding a comment explaining the situation? (The docstring is more for external use.)

ilevkivskyi commented:
@gvanrossum Thank you for another review! I implemented requested changes and added comments in code and one small test.

gvanrossum commented:
Whee! Hope @davidfstr can now start implementing some of the features implemented here. (We should probably kill some `# type: ignore` comments once that's done.)

ilevkivskyi commented:
I think we could split the work here. I could take care about class based syntax in Python 3.6, while @davidfstr could take care about keyword based syntax https://github.com/python/mypy/issues/2492

davidfstr commented:
Yep sounds good to me. I was actually planning to work on #2492 next.
